EIPS/eip-2315.md

```diff
@@ -0,0 +1,91 @@
+---
+eip: 2315
+title: Simple Subroutines for the EVM
+status: Draft
+type: Standards Track
+category: Core
+author: Greg Colvin
+discussions-to: https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm/3941
+created: 2019-10-17
+---
+## Abstract
+
+This proposal introduces two opcodes to support subroutines:  `JUMPSUB` and `RETSUB`.
+
+## Motivation
+
+The EVM does not provide subroutines as a primitive.  Instead, calls must be synthesized by fetching and pushing the current program counter on the data stack and jumping to the subroutine address; returns must be synthesized by getting the return address to the top of stack and jumping back to it.
+
+## Specification
+
+**_JUMPSUB_**
+
+* Jumps to the address on top of the stack.  This must be a `JUMPDEST`.
+
+**_RETSUB_**
+
+* Returns to the most recently executed `JUMPSUB` instruction.
+
+## Rationale
+
+This is the smallest possible change that provides native subroutines without breaking backwards compatibility.  It does not enforce validity like [EIP-615](https://eips.ethereum.org/EIPS/eip-615), but code that
+* pushes a constant argument before every `JUMP`, `JUMPI` and `JUMPSUB`
+* ensures that the _data stack_ is the same size at every execution of each `JUMP`, `JUMPI`, `JUMPSUB` or `JUMPDEST` operation.
+
+is valid and will meet all of the [EIP-615](https://eips.ethereum.org/EIPS/eip-615) safety conditions.  In particular, the Yellow paper defines five exceptional halting states.  
+1. Insufficient gas
+2. More than 1024 stack items
+3. Insufficient stack items
+4. Invalid jump destination
+5. Invalid instruction
+
+Valid code cannot violate states 3, 4, or 5.
```

Comment:
Not quite true, as written. 

- It can violate `4` if the `constant` being pushed is incorrect, 
- It can violate `3` if the code is simply `POP` (valid according to the rules above), 
- Same for bullet 5. 

Maybe I totally misread what you meant. I interpret it as 

> Code that adheres to these two rules is 'valid'. Code that is 'valid' can never violate 3,4 or 5. 

Maybe you meant 

> Code that adheres to these two rules and never violates 3,4,5 is `valid`

Reply:
I'm yanking this.